README.md
# safari.rs

safari.rs provides some tools for interacting with Safari on the command-line.

## Commands

1.  Get the URL from a given window or tab:

    ```console
    $ # Get the URL of the frontmost tab of the frontmost window
    $ safari url
    https://github.com

    $ # Get a URL from the second window from the tab
    $ safari url --window=2
    https://example.com/foo

    $ # Third tab from the left of the second window
    $ safari url --window=2 --tab=3
    https://example.com/foo
    ```

    I have the first two commands bound to shortcuts `furl` and `2url` for quick access.

2.  Get a list of URLs from every open tab:

    ```console
    $ safari list-tabs
    https://github.com
    https://example.com/foo
    https://crates.io/crates/urlparse
    ...
    ```

3.  Go through and batch close tabs:

    ```console
    $ safari clean-tabs youtube.com,twitter.com
    ```

    I find this useful for quickly cutting down my open tabs.

4.  Get a list of URLs from Reading List:

    ```console
    $ safari reading-list
    ```

5.  Get a list of URLs from all your devices with iCloud Tabs:

    ```console
    $ safari icloud-tabs
    ```

    You can get a list of known devices with `icloud-tabs --list-devices`, and filter the URLs with the `--device` flag:

    ```console
    $ safari icloud-tabs --device="Alex's iPhone"
    ```

## Installation

You need [Rust installed][rust].
Then to install:

```console
$ cargo install --git https://github.com/alexwlchan/safari.rs
```

It's tested in Travis with the current version of stable Rust, but it only gets tested when it was last modified (at time of writing, that was Rust 1.40.0).

[rust]: https://www.rust-lang.org/en-US/install.html

## URL transformations

The commands that produce URLs do a bit of cleaning before they return:

*   Convert links for a mobile site to the desktop site – for example, converting `mobile.twitter.com` to `twitter.com`.
*   Stripping tracking junk (e.g. UTM tracking parameters) from URLs.
*   Removing some extraneous information that isn't generally useful.

## Motivation

I first got the idea for a script to access Safari URLs [from Dr. Drang][dr].
I've been through several different versions – AppleScript, shell, Python –
gradually adding the cleaning features – and now I've written a new
version in Rust.

Why Rust?

*   It's really fast.  The Rust script returns immediately – when I tried writing this with Python, I had a noticeable delay when typing `;furl`.
    This is a tool I use dozens of times a week, so every second counts.
*   I like Rust, and I’ve been enjoying playing with it recently.

[dr]: http://www.leancrew.com/all-this/2009/07/safari-tab-urls-via-textexpander/

## License

MIT.


CHANGELOG.md
# Changelog

## v2.3.14 (2023-12-17)

*   URL tweak: remove tracking parameters from Etsy URLs.

## v2.3.12 (2023-02-20)

*   URL tweak: remove another parameter from Twitter URLs.

## v2.3.11 (2023-02-07)

*   URL tweak: remove more referrer parameters from Twitter URLs.

## v2.3.10 (2023-01-02)

*   URL tweak: remove HubSpot tracking parameters.
*   Provide binaries for Macs running Apple Silicon.

## v2.3.9 (2022-05-14)

*   Start publishing binaries as GitHub Releases.

## v2.3.8 (2022-04-17)

*   URL tweaking: remove the `frs` parameter from URLs on Etsy.

## v2.3.7 (2021-10-26)

*   URL tweak: remove the same tracking parameters from Amazon Smile URLs as regular Amazon URLs.

## v2.3.6 (2021-09-08)

*   URL tweak: remove all tracking parameters from Tiktok URLs.

## v2.3.5 (2021-08-08)

*   URL tweak: remove Cloudflare query parameters that start `__cf`.

## v2.3.4 (2021-02-05)

*   URL tweak: strip the `pro` parameter from URLs on Etsy.

## v2.3.3 (2021-02-05)

*   URL tweak: strip the share parameter from URLs on <twitter.com>.

## v2.3.2 (2020-11-21)

*   URL tweak: strip a tracking parameter from URLs on <wordery.com>.

## v2.3.1 (2020-10-02)

*   URL tweak: strip the `source=` parameter from URLs on <stacks.wellcomecollection.org>.

## v2.3.0 (2020-07-26)

*   Add a new command `title` to get the title of a Safari window.

## v2.2.9 (2020-07-26)

*   Internal refactoring to remove special-case handling for the old Wellcome Images site (`wellcomeimages.org`).
    Since that site was shut down over two years ago, the code to handle it can be removed without a user-facing effect.

## v2.2.8 (2020-02-09)

*   URL tweak: strip the `ref=` and `sort=` parameters from Redbubble URLs.

## v2.2.7 (2020-01-12)

*   URL tweak: strip the `m=` parameter from Blogspot URLs (so links are never mobile links).

## v2.2.6 (2019-12-29)

*   URL tweak: strip the `app=` parameter from YouTube URLs.

## v2.2.5 (2019-12-28)

*   URL tweak: strip more tracking parameters from Etsy URLs.

## v2.2.4 (2019-04-21)

*   URL tweak: strip tracking parameters from Etsy URLs.

## v2.2.3 (2018-05-22)

*   URL tweak: links to the files tab of GitHub pull requests are now replaced
    by links to the top of the pull request.

## v2.2.2 (2017-10-07)

*   URL tweak: mobile.nytimes.com links are now replaced by non-mobile versions.
    ([#59](https://github.com/alexwlchan/safari.rs/issues/59))

## v2.2.1 (2017-08-29)

*   URL tweak: Remove the `_ga` tracking parameter from shared URLs.
    ([#56](https://github.com/alexwlchan/safari.rs/issues/56))

## v2.2.0 (2017-06-10)

*   New command: the `resolve` command can take a URL as an argument, and print to stdout the final location of that URL after following any redirects.
    Useful for working with, e.g., `t.co` or `bit.ly` URLs.

## v2.1.2 (2017-06-10)

*   URL tweak: remove tracking parameters from URLs on `telegraph.co.uk`.
    ([#48](https://github.com/alexwlchan/safari.rs/issues/48))

## v2.1.1 (2017-06-05)

*   URL tweak: most of `wellcomeimages.org` is loaded entirely in `<iframe>` tags, and the page URL doesn't change.
    This isn't very useful, so now the tool will attempt to guess a sensible permalink on image pages, based on the HTML.

## v2.1.0 (2017-06-03)

*   Deprecation: the `urls-all` command has been replaced by `list-tabs`.
    This reverts the change in v2.0, because I kept trying to use the old command, and realised I actually prefer the old command.
    ([#41](https://github.com/alexwlchan/safari.rs/pull/41))
*   New command: the `tidy-url` command can take a URL as an argument, and
    pass it through the URL cleaning pipeline.
    ([#43](https://github.com/alexwlchan/safari.rs/pull/43))

## v2.0.7 (2017-06-02)

*   URL tweak: all stackexchange.com URLs now use the referral links.

## v2.0.6 (2017-05-29)

*   URL tweak: scifi.stackexchange.com URLs now uses the Share links provided for SFF.SE referrals.

## v2.0.5 (2017-05-18)

*   Bugfix: don't include `://` in the list of currently open Safari tabs.
    ([#35](https://github.com/alexwlchan/safari.rs/issues/35))

## v2.0.4 (2017-05-08)

*   URL tweak: Stack Overflow URLs now use the Share links provided for SO referrals.

## v2.0.3 (2017-05-06)

*   URL tweak: discard `?highlight` and fragment from `docs.python.org` URLs
    if they're pointing at a module on a module page.

## v2.0.2 (2017-05-02)

*   Bugfix: don't panic on the `icloud-tabs` command if there's a device that doesn't have any tabs open ([#23](https://github.com/alexwlchan/safari.rs/issues/23)).

## v2.0.1 (2017-04-30)

*   Bugfix: don't include `://missing value` in the list of currently open Safari tabs.

## v2.0.0 (2017-04-30)

*   A near-complete rewrite that overhauls the command-line parsing, improves the interface, and has better error reporting.
*   Replace the `furl` and `2url` commands with a more generic `url` command with configurable `--window` and `--tab` parameters.
*   Rename the `list-tabs` command to `urls-all`.
*   Add a new `icloud-tabs` command for getting data about iCloud Tabs.
*   Strip UTM tracking parameters from all URLs.
*   Bugfix: don't panic when processing a YouTube URL that isn't a video.

## v1.2.0 (2017-02-25)

*   Add a `reading-list` command that a list of URLs from Reading List.

## v1.1.0 (2017-02-23)

*   Renamed from `safari-url` to `safari`.
*   Add a `clean-tabs` command that closes tabs whose URLs match a particular pattern.
*   Add a `list-tabs` command that gets a list of URLs in every open tab.
*   Strip tracking data from Buzzfeed and Mashable URLs.

## v1.0.0 (2017-02-20)

*   Initial release (under the name `safari-url`).


test_output.py
#!/usr/bin/env python
# -*- encoding: utf-8
"""
These are tests of the external behaviour -- feature tests, if you like.
They run the compiled binaries, and make assertions about the return code,
stdout and stderr.
"""

import unittest

from conftest import BaseTest


class TestSafariRS(BaseTest):
    def test_urls_all_flag_is_deprecated(self):
        result = self.run_safari_rs("urls-all")
        self.assertIn("deprecated", result.stderr)

    def test_list_tabs_flag_is_not_deprecated(self):
        result = self.run_safari_rs("list-tabs")
        self.assertNotIn("deprecated", result.stderr)

    def test_no_extra_whitespace_on_tidy_url(self):
        result = self.run_safari_rs(
            "tidy-url", "https://github.com/alexwlchan/safari.rs/issues"
        )
        assert result.rc == 0
        assert result.stderr == ""
        assert result.stdout.strip() == result.stdout

    def _assert_resolve_tco(self, url, expected):
        result = self.run_safari_rs("resolve", url)
        assert result.rc == 0
        assert result.stderr == ""
        assert result.stdout == expected

    def test_resolve_single_redirect(self):
        self._assert_resolve_tco(
            "https://t.co/2pciHpqpwC",
            "https://donmelton.com/2013/06/04/remembering-penny/",
        )

    def test_resolve_multiple_redirect(self):
        self._assert_resolve_tco(
            "https://t.co/oSJaiNlIP6", "https://bitly.com/blog/backlinking-strategy/"
        )

    def test_resolve_no_redirect(self):
        self._assert_resolve_tco("https://example.org/", "https://example.org/")


if __name__ == "__main__":
    unittest.main()


conftest.py
#!/usr/bin/env python
# -*- encoding: utf-8

import collections
import os
import subprocess
import unittest


Result = collections.namedtuple('Result', 'rc stdout stderr')

ROOT = subprocess.check_output([
    'git', 'rev-parse', '--show-toplevel']).decode('ascii').strip()

BINARY = os.path.join(ROOT, 'target', 'release', 'safari')


subprocess.check_call(['cargo', 'build', '--release'], cwd=ROOT)


class BaseTest(unittest.TestCase):

    def run_safari_rs(self, *args):
        proc = subprocess.Popen([BINARY] + list(args),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE
        )
        stdout, stderr = proc.communicate()
        return Result(
            rc=proc.returncode,
            stdout=stdout.decode('ascii'),
            stderr=stderr.decode('ascii')
        )


src/main.rs
#![deny(warnings)]

extern crate dirs;
extern crate docopt;
extern crate plist;
extern crate reqwest;
#[macro_use]
extern crate serde_derive;
extern crate tera;
extern crate urlencoding;
extern crate urlparse;

use std::io::Write;
use std::process;

mod applescript;
mod cli;
mod safari;
mod urls;

const NAME: &str = env!("CARGO_PKG_NAME");
const VERSION: &str = env!("CARGO_PKG_VERSION");

// http://stackoverflow.com/a/27590832/1558022
macro_rules! error(
  ($($arg:tt)*) => {
    {
      let r = writeln!(&mut ::std::io::stderr(), $($arg)*);
      r.expect("failed printing to stderr");
      process::exit(1);
    }
  }
);

/// Exits the program if Safari isn't running.
fn assert_safari_is_running() {
    if !safari::is_safari_running() {
        error!("Safari is not running.");
    }
}

fn main() {
    let args = cli::parse_args(NAME);

    if args.flag_version {
        println!("{}.rs v{}", NAME, VERSION);
    }

    if args.cmd_url {
        assert_safari_is_running();
        match safari::get_url(args.flag_window, args.flag_tab) {
            Ok(url) => print!("{}", url),
            Err(e) => error!("{}", e),
        };
    }

    if args.cmd_title {
        assert_safari_is_running();
        match safari::get_title(args.flag_window, args.flag_tab) {
            Ok(url) => print!("{}", url),
            Err(e) => error!("{}", e),
        };
    }

    if args.cmd_resolve {
        print!("{}", urls::resolve(&args.arg_url));
    }

    if args.cmd_list_tabs {
        assert_safari_is_running();
        for url in safari::get_all_urls() {
            println!("{}", url);
        }
    }

    if args.cmd_close_tabs {
        assert_safari_is_running();
        let patterns = args.arg_urls_to_close.split(",").collect();
        safari::close_tabs(patterns);
    }

    if args.cmd_reading_list {
        match safari::get_reading_list_urls() {
            Ok(urls) => {
                for url in urls {
                    println!("{}", url);
                }
            }
            Err(e) => error!("{}", e),
        };
    }

    if args.cmd_icloud_tabs {
        if args.flag_list_devices {
            match safari::list_icloud_tabs_devices() {
                Ok(devices) => {
                    for device in devices {
                        println!("{}", device);
                    }
                }
                Err(e) => error!("{}", e),
            };
        } else {
            let tab_data = match safari::get_icloud_tabs_urls() {
                Ok(tab_data) => tab_data,
                Err(e) => error!("{}", e),
            };
            match args.flag_device {
                Some(d) => match tab_data.get(&d) {
                    Some(urls) => {
                        for url in urls {
                            println!("{}", url);
                        }
                    }
                    None => (),
                },
                None => {
                    for urls in tab_data.values() {
                        for url in urls {
                            println!("{}", url);
                        }
                    }
                }
            }
        }
    }

    if args.cmd_tidy_url {
        print!("{}", urls::tidy_url(&args.arg_url));
    }
}


src/safari.rs
use std::collections::{BTreeMap, HashMap};
use std::fs::File;
use std::process;

use plist::Plist;
use tera::{Context, Tera};

use applescript::run as run_applescript;
use urls;

macro_rules! error(
  ($($arg:tt)*) => { { return Err(format!($($arg)*)) } }
);

/// Returns true/false if Safari is running.
pub fn is_safari_running() -> bool {
    let cmd_result = process::Command::new("ps")
        .arg("-eaf")
        .output()
        .expect("Unable to test if Safari is running.");
    for line in String::from_utf8(cmd_result.stdout).unwrap().lines() {
        if line.contains("Safari.app/Contents/MacOS/Safari") {
            return true;
        }
    }
    false
}

/// Return a URL from a Safari window.
///
/// Given a (window, tab) pair, this function looks up the URL of the tab.
/// Note that it doesn't do any error handling, so will throw an execution
/// error if it fails.
///
/// * `window` - Window index.  1 is frontmost.  If None, assumes the
///              frontmost window.
/// * `tab` - Tab index.  1 is leftmost.  If None, assumes the frontmost tab.
///
pub fn get_url(window: Option<u32>, tab: Option<u32>) -> Result<String, String> {
    match get_property(window, tab, "URL") {
        Ok(url) => Ok(urls::tidy_url(&url)),
        Err(e) => Err(e),
    }
}

/// Return a title from a Safari window.
///
/// Given a (window, tab) pair, this function looks up the URL of the tab.
/// Note that it doesn't do any error handling, so will throw an execution
/// error if it fails.
///
/// * `window` - Window index.  1 is frontmost.  If None, assumes the
///              frontmost window.
/// * `tab` - Tab index.  1 is leftmost.  If None, assumes the frontmost tab.
///
pub fn get_title(window: Option<u32>, tab: Option<u32>) -> Result<String, String> {
    get_property(window, tab, "name")
}

/// Look up a property on a Safari window.
///
/// Given a (window, tab) pair, this function looks up the property of that tab.
/// Note that it doesn't do any error handling, so will throw an execution
/// error if it fails.
///
/// * `window` - Window index.  1 is frontmost.  If None, assumes the
///              frontmost window.
/// * `tab` - Tab index.  1 is leftmost.  If None, assumes the frontmost tab.
/// * `property` - Name of the property, as defined in the OSA scripting dictionary.
///
fn get_property(window: Option<u32>, tab: Option<u32>, property: &str) -> Result<String, String> {
    // If a tab isn't specified, assume the user wants the frontmost tab.
    let command = match window {
        Some(w_idx) => match tab {
            Some(t_idx) => format!(
                "tell application \"Safari\" to get {} of tab {} of window {}",
                property, t_idx, w_idx
            ),
            None => format!(
                "tell application \"Safari\" to get {} of document {}",
                property, w_idx
            ),
        },
        None => format!(
            "tell application \"Safari\" to get {} of document 1",
            property
        ),
    };
    let output = run_applescript(&command);

    if output.status.success() {
        Ok(output.stdout.trim().to_owned())
    } else {
        if output.stderr.contains("Invalid index") {
            error!("Invalid index: no such window or tab.")
        } else {
            error!("Unexpected error from osascript: {:?}", output.stderr)
        }
    }
}

/// Tests for get_property().
#[cfg(test)]
mod tests_property {
    use safari::get_property;

    #[test]
    fn test_invalid_property_is_rejected() {
        let result = get_property(None, None, "fooble");
        assert!(result.is_err());
    }
}

/// Return a list of URLs from Safari.
///
/// This returns a list of URLs, one for every tab in Safari.  Iteration
/// order depends on AppleScript, which I don't think is guaranteed to be
/// stable (in particular, I think it depends on which window is frontmost).
///
pub fn get_all_urls() -> Vec<String> {
    let windows = get_window_tab_count_pairs();
    let mut urls = vec![];
    for window in windows {
        for tab in 1..window.tab_count {
            match get_url(Some(window.window_index), Some(tab)) {
                Ok(u) => urls.push(u),
                Err(_) => {}
            }
        }
    }
    urls
}

/// Convert URL patterns into AppleScript conditions.
///
/// These conditions can be used in an `if` statement in AppleScript to
/// decide whether a tab should be closed.  Three patterns are supported,
/// a limited regex syntax:
///
///     example.com             matches anywhere in the URL
///     ^http://examples.com    matches at the start of the URL
///     example.com/$           matches at the end of the URL
///
fn parse_conditions(url_patterns: Vec<&str>) -> Vec<String> {
    url_patterns
        .iter()
        .map(|p| {
            if p.starts_with("^") {
                format!("starts with \"{}\"", p.replace("^", ""))
            } else if p.ends_with("$") {
                format!("ends with \"{}\"", p.replace("$", ""))
            } else {
                format!("contains \"{}\"", p)
            }
        })
        .collect()
}

/// Tests for parse_conditions().
#[cfg(test)]
mod tests {
    use safari::parse_conditions;

    #[test]
    fn test_parse_conditions() {
        let patterns = vec!["github.com", "^facebook.com", "twitter.com$"];
        let expected = vec![
            "contains \"github.com\"",
            "starts with \"facebook.com\"",
            "ends with \"twitter.com\"",
        ];
        let actual = parse_conditions(patterns);
        assert_eq!(actual, expected);
    }
}

/// Close tabs in Safari that match URL patterns.
///
/// Takes a list of URL patterns, and tries to close any matching Safari tabs.
/// Bugginess in AppleScript means this isn't always perfect, but we can
/// have a go.  In general it will fail to close tabs, rather than close
/// the wrong tabs.
///
pub fn close_tabs(url_patterns: Vec<&str>) {
    let conditions = parse_conditions(url_patterns);

    let clean_tabs_template = include_str!("scripts/clean-tabs.scpt");
    let mut context = Context::new();
    context.add("conditions", &conditions);
    let script = Tera::one_off(&clean_tabs_template, &context, false).unwrap();

    // Run it twice to get around weird AppleScript bugs.
    run_applescript(&script);
    run_applescript(&script);
}

/// Get the Bookmarks.plist dict for a given title
fn read_bookmarks_plist(title: &str) -> Result<Plist, String> {
    // All Safari data lives at ~/Library/Safari/Bookmarks.plist
    // TODO: There's probably a more idiomatic Rust-like way to write this.
    let mut plist_path = match dirs::home_dir() {
        Some(v) => v,
        None => error!("Unable to get home directory?"),
    };
    plist_path.push("Library/Safari/Bookmarks.plist");

    let file = match File::open(plist_path) {
        Ok(v) => v,
        Err(e) => error!("Unable to open ~/Library/Safari/Bookmarks.plist: {:?}", e),
    };

    let plist = match Plist::read(file) {
        Ok(v) => v,
        Err(e) => error!("Unable to read ~/Library/Safari/Bookmarks.plist: {:?}", e),
    };

    let data = match plist.as_dictionary() {
        Some(v) => v,
        None => error!("Unable to parse ~/Library/Safari/Bookmarks.plist as dictionary?"),
    };

    // The structure of Bookmarks.plist is as follows:
    //
    //     <dict>
    //       <key>Children</key>
    //       <array>
    //         <dict>
    //           <key>Title</key><string>History</string>
    //           ... dict data ...
    //         </dict>
    //         <dict>
    //           <key>Title</key><string>com.apple.ReadingList</string>
    //           ... dict data ...
    //         </dict>
    //         ... other array items ...
    //       </array>
    //     </dict>
    //
    let children = match data.get("Children") {
        Some(child_key) => match child_key.as_array() {
            Some(v) => v,
            None => {
                error!("Top-level children key in ~/Library/Safari/Bookmarks.plist isn't an array?")
            }
        },
        None => error!("Unable to find top-level Children key in ~/Library/Safari/Bookmarks.plist"),
    };

    let mut matching_children = children.iter().filter(|d| match d.as_dictionary() {
        Some(dict) => match dict.get("Title") {
            Some(title_elem) => match title_elem.as_string() {
                Some(v) => v == title,
                None => false,
            },
            None => false,
        },
        None => false,
    });

    // Check we got one, and only one result.
    let result = match matching_children.next() {
        Some(v) => v,
        None => error!("Unable to find key {} in Bookmarks.plist", title),
    };

    match matching_children.next() {
        Some(_) => error!("Got more than one result for {} in Bookmarks.plist", title),
        None => {}
    };

    Ok(result.to_owned())
}

/// Return a list of URLs from Reading List.
///
/// Iteration order depends on the order in which they're stored in
/// Bookmarks.plist, which is usually (but not guaranteed to be) newest first.
///
pub fn get_reading_list_urls() -> Result<Vec<String>, String> {
    let plist = match read_bookmarks_plist("com.apple.ReadingList") {
        Ok(v) => v,
        Err(e) => return Err(e),
    };

    // TODO: All these unwrap() calls should probably be handled better
    let children = plist
        .as_dictionary()
        .unwrap()
        .get("Children")
        .unwrap()
        .as_array()
        .unwrap();

    Ok(children
        .iter()
        .map(|child| {
            child
                .as_dictionary()
                .unwrap()
                .get("URLString")
                .unwrap()
                .as_string()
                .unwrap()
        })
        .map(|url| urls::tidy_url(url))
        .collect())
}

/// Get the com.apple.Safari.plist preferences file
fn read_safari_plist() -> Result<BTreeMap<String, Plist>, String> {
    // All Safari data lives at ~/Library/Safari/Bookmarks.plist
    // TODO: There's probably a more idiomatic Rust-like way to write this.
    let mut plist_path = match dirs::home_dir() {
        Some(v) => v,
        None => error!("Unable to get home directory?"),
    };
    plist_path.push("Library/SyncedPreferences/com.apple.Safari.plist");

    let file = match File::open(plist_path) {
        Ok(v) => v,
        Err(e) => error!("Unable to open com.apple.Safari.plist: {:?}", e),
    };

    let plist = match Plist::read(file) {
        Ok(v) => v,
        Err(e) => error!("Unable to read com.apple.Safari.plist: {:?}", e),
    };

    let data = match plist.as_dictionary() {
        Some(v) => v,
        None => error!("Unable to parse com.apple.Safari.plist as dictionary?"),
    };

    // The structure of com.apple.Safari.plist is as follows:
    //
    //      <dict>
    //        <key>values</key>
    //        <dict>
    //          <dict>
    //            <key>[[ UUID ]]</key>
    //            <dict>
    //              ... dict data ...
    //              <key>value</key>
    //              <dict>
    //                ... dict data ...
    //                <key>DeviceName</key><string>[[ Device name ]]</string>
    //                <key>Tabs</key>
    //                <array>
    //                  <dict>
    //                    <key>Title</key><string>[[ Title ]]</string>
    //                    <key>URL</key><string>[[ URL ]]</string>
    //                  </dict>
    //                  ... array items for other tabs ...
    //                </array>
    //              </dict>
    //            </dict>
    //            ... dicts for other devices ...
    //          </dict>
    //        </dict>
    //      </dict>
    //
    match data.get("values") {
        Some(child_key) => match child_key.as_dictionary() {
            Some(v) => Ok(v.to_owned()),
            None => error!("Top-level values key in com.apple.Safari.plist isn't an dictionary?"),
        },
        None => error!("Unable to find top-level values key in com.apple.Safari.plist"),
    }
}

/// Return a list of devices in iCloud Tabs.
pub fn list_icloud_tabs_devices() -> Result<Vec<String>, String> {
    let plist = match read_safari_plist() {
        Ok(plist) => plist,
        Err(e) => return Err(e),
    };

    Ok(plist
        .values()
        .map(|value| {
            value
                .as_dictionary()
                .unwrap()
                .get("value")
                .unwrap()
                .as_dictionary()
                .unwrap()
                .get("DeviceName")
                .unwrap()
                .as_string()
                .unwrap()
                .to_owned()
        })
        .collect())
}

/// Return a list of URLs from iCloud Tabs.
pub fn get_icloud_tabs_urls() -> Result<HashMap<String, Vec<String>>, String> {
    let plist = match read_safari_plist() {
        Ok(plist) => plist,
        Err(e) => return Err(e),
    };

    // Within the `values` dictionary, each device is as follows:
    //
    //    <key>[[ device UUID ]]</key>
    //    <dict>
    //      [[ device data ]]
    //      <dict>
    //        <key>Tabs</key>
    //        <key>DeviceName</key><string>[[ device name ]]</key>
    //        <key>Tabs</key>
    //        <array>
    //          <dict>
    //            <key>URL</key><string>[[ URL ]]</string>
    //            [[ other tab data ]]
    //          </dict>
    //          [[ other tabs ]]
    //
    let mut result: HashMap<String, Vec<String>> = HashMap::new();
    let devices_with_tabs = plist.values().map(|value| {
        value
            .as_dictionary()
            .unwrap()
            .get("value")
            .unwrap()
            .as_dictionary()
            .unwrap()
    });
    for data in devices_with_tabs {
        let name: String = data
            .get("DeviceName")
            .unwrap()
            .as_string()
            .unwrap()
            .to_owned();

        // If a device is registered with iCloud but Safari isn't running or
        // there aren't any tabs open, there isn't a Tabs field.
        if !data.contains_key("Tabs") {
            continue;
        }

        let urls: Vec<String> = data
            .get("Tabs")
            .unwrap()
            .as_array()
            .unwrap()
            .iter()
            .map(|tab| {
                tab.as_dictionary()
                    .unwrap()
                    .get("URL")
                    .unwrap()
                    .as_string()
                    .unwrap()
            })
            .map(|url| urls::tidy_url(url))
            .collect();
        result.insert(name, urls);
    }
    Ok(result)
}

struct SafariWindow {
    window_index: u32,
    tab_count: u32,
}

/// Get a list of (window, tab_count) pairs for the currently running
/// version of Safari.
///
/// There are lots of bugs in Safari's AppleScript handler, so knowing
/// that there are N windows does not imply you can look up the tabs for
/// each window 1, ..., N.  There might be gaps -- looking up a window
/// in the middle could crash the AppleScript handler.
fn get_window_tab_count_pairs() -> Vec<SafariWindow> {
    let mut pairs = vec![];
    let r = run_applescript("tell application \"Safari\" to get count of windows");
    let window_count = r.stdout.trim().parse::<u32>().unwrap();
    for window in 1..(window_count + 1) {
        let r = run_applescript(&format!(
            "tell application \"Safari\" to get count of tabs of window {}",
            window
        ));
        if r.status.success() {
            pairs.push(SafariWindow {
                window_index: window,
                tab_count: r.stdout.trim().parse::<u32>().unwrap(),
            });
        }
    }
    pairs
}


se_referral_autogen.py
#!/usr/bin/env python
# -*- encoding: utf-8
"""
This script auto-generates the ``fix_se_referral`` lines used in ``urls.rs``.

It takes somebody's Stack Exchange account URL, of the form

    https://stackexchange.com/users/:user_id/:user_name

and spits out a list of ``fix_se_referral`` calls.
"""

import sys
from urllib.parse import urlparse

import bs4
import requests


se_account_url = sys.argv[1]
resp = requests.get(se_account_url, params={'tab': 'accounts'})

soup = bs4.BeautifulSoup(resp.text, 'html.parser')
accounts = []
for account in soup.find_all('div', attrs={'class': 'account-container'}):
    user_page_url = account.find('h2').find('a').attrs['href']
    components = urlparse(user_page_url)
    user_id = components.path.split('/')[2]
    accounts.append((components.netloc, user_id))

for netloc, user_id in sorted(accounts):
    print(f'fix_se_referral(&mut parsed_url, "{netloc}", "{user_id}");')


src/cli.rs
use std::io::Write;

use docopt::{Docopt, Error};

// https://stackoverflow.com/a/27590832/1558022
macro_rules! println_stderr(
    ($($arg:tt)*) => { {
        let r = writeln!(&mut ::std::io::stderr(), $($arg)*);
        r.expect("failed printing to stderr");
    } }
);

const USAGE: &str = "
Usage: <NAME> url [--window=<WINDOW> [--tab=<TAB>]]
       <NAME> title [--window=<WINDOW> [--tab=<TAB>]]
       <NAME> tidy-url <url>
       <NAME> resolve <url>
       <NAME> list-tabs
       <NAME> urls-all
       <NAME> close-tabs <urls-to-close>
       <NAME> reading-list
       <NAME> icloud-tabs [--list-devices | --device=<DEVICE>]
       <NAME> (-h | --help)
       <NAME> --version

Options:
    -h --help           Show this screen.
    --version           Show version.
    --window=<WINDOW>   Which window to choose a URL from.  Use 1 for the
                        frontmost window, 2 for the second window, and so on.
    --tab=<TAB>         Which tab to choose a URL from.  Use 1 for the leftmost
                        tab, 2 for second-from-left, and so on.
    --list-devices      Get a list of all the devices known to iCloud Tabs.
    --device=<DEVICE>   Only get iCloud URLs for this device.

Commands:
    url           Print a URL from an open Safari tab.
    title         Print the title of an open Safari tab.
    resolve       Follow redirects and print the final location of a URL.
    tidy-url      Remove tracking junk, mobile, links, etc. from a URL.
    list-tabs     Prints a list of URLs from every open Safari tab.
    urls-all      Same as urls-all.  Deprecated.
    close-tabs    Close any tabs with the given URLs.
    reading-list  Print a list of URLs from Reading List.
    icloud-tabs   Get a list of URLs from iCloud Tabs.  Default is to list URLs
                  from every device, or you can filter with the --device flag.
";

#[derive(Debug, Deserialize)]
pub struct Args {
    pub cmd_url: bool,
    pub cmd_title: bool,
    pub cmd_tidy_url: bool,
    pub cmd_resolve: bool,
    pub cmd_urls_all: bool,
    pub cmd_list_tabs: bool,
    pub cmd_close_tabs: bool,
    pub cmd_icloud_tabs: bool,
    pub cmd_reading_list: bool,
    pub flag_window: Option<u32>,
    pub flag_tab: Option<u32>,
    pub flag_version: bool,
    pub flag_list_devices: bool,
    pub flag_device: Option<String>,
    pub arg_url: String,
    pub arg_urls_to_close: String,
}

pub fn parse_args(name: &str) -> Args {
    let mut args: Args = Docopt::new(str::replace(USAGE, "<NAME>", name))
        .and_then(|d| d.deserialize())
        .unwrap_or_else(|e| e.exit());

    // 0 is the default value for the --window and --tab flags, so if we get
    // this value then replace it with None.
    if args.cmd_url {
        match args.flag_window {
            Some(v) => {
                if v == 0 {
                    args.flag_window = None;
                };
            }
            None => {}
        };
        match args.flag_tab {
            Some(v) => {
                if v == 0 {
                    args.flag_tab = None;
                };
            }
            None => {}
        };

        if args.flag_tab.is_some() && args.flag_window.is_none() {
            Error::Usage("Cannot use --tab without --window.".to_string()).exit();
        }
    }

    if args.cmd_urls_all {
        println_stderr!("The --urls-all flag is deprecated; please use --list-tabs.");
        args.cmd_urls_all = false;
        args.cmd_list_tabs = true;
    }

    args
}
